crates/wasmparser/src/validator/names.rs

```diff
@@ -320,12 +345,66 @@ impl KebabName {
                         bail!(offset, "failed to parse version: {e}")
                     }
                 }
-                ParsedKebabName::Id {
+                ParsedKebabName::InterfaceId {
                     colon: colon as u32,
                     slash: slash as u32,
                     at: at.map(|i| i as u32),
                 }
             }
+            ComponentExternName::Implementation(s) => match s {
+                crate::ImplementationImport::Url(metadata) => {
+                    validate_kebab(s.as_str())?;
+                    ParsedKebabName::Normal
+                }
+                crate::ImplementationImport::Relative(metadata) => {
+                    validate_kebab(s.as_str())?;
+                    ParsedKebabName::Normal
+                }
+                crate::ImplementationImport::Naked(metadata) => {
+                    validate_kebab(s.as_str())?;
+                    ParsedKebabName::Normal
+                }
+                crate::ImplementationImport::Locked(metadata) => {
+                    let mut colon_acc = Vec::new();
+                    let colon = find_many(metadata.as_str(), ':', &mut colon_acc);
+                    if colon.len() == 0 {
+                        bail!(offset, "At least one namespace in registry id is required");
+                    }
+                    let mut slash_acc = Vec::new();
+                    let slash = find_many(
+                        &metadata.as_str()[colon[colon.len() - 1]..],
+                        '/',
+                        &mut slash_acc,
+                    );
+                    let last_slash = slash[slash.len() - 1];
+                    let at = s.as_str()[last_slash..].find('@').map(|i| i + last_slash);
+                    ParsedKebabName::RegistryId {
+                        colon: colon.iter().map(|i| *i as u32).collect::<Vec<u32>>(),
+                        slash: colon.iter().map(|i| *i as u32).collect::<Vec<u32>>(),
+                        at: at.map(|i| i as u32),
+                    }
+                }
```

Comment:
This one is unfortunately pretty tricky. I don't think the tooling is really ready to support the generalization of ids landed in https://github.com/WebAssembly/component-model/pull/222. There's a lot of work to do on the WIT side of things to support that which isn't part of this PR and would eventually need to make its way in there. For example the `(interface "...")` production should be using the same validation that's added here for `locked`, but it's not currently.

I think it would be best for now to update `RegistryId` to specify whether the `at` is either "none", a "semver", or a "semver range". Otherwise though the locked/unlocked/interface productions here should all use `RegistryId` in the same format, which I think for now we'll need to limit to `foo:bar/baz` with no more colons or slashes. I realize that's not up-to-date with the spec but there will need to be future work for that generalization which is plumbed through to WIT everywhere too.